### PR TITLE
dtools dmd 2.106.1

### PR DESCRIPTION
**Pull request by timsutton**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by p-linnane** [comment]
@timsutton Please update the `phobos` resource as indicated in the comment and test failure.

**Comment by timsutton** [comment]
> @timsutton Please update the `phobos` resource as indicated in the comment and test failure.

Done! Sorry, had missed that earlier.

**Comment by timsutton** [comment]
```
  ==> Testing dmd
  ==> /usr/local/Cellar/dmd/2.106.1/bin/dmd -fPIC /usr/local/Cellar/dmd/2.106.1/share/dmd/samples/hello.d
  ld: multiple errors: symbol count from symbol table and dynamic symbol table differ in '/private/tmp/dmd-test-20240102-11622-4jskt3/hello.o' in '/private/tmp/dmd-test-20240102-11622-4jskt3/hello.o'; address=0x0 points to section(2) with no content in '/usr/local/opt/dmd/lib/libphobos2.a[3228](config_a94_4c3.o)'
  clang: error: linker command failed with exit code 1 (use -v to see invocation)
  Error: linker exited with status 1
  Error: dmd: failed
  An exception occurred within a child process:
    BuildError: Failed executing: /usr/local/Cellar/dmd/2.106.1/bin/dmd -fPIC /usr/local/Cellar/dmd/2.106.1/share/dmd/samples/hello.d
```

**Review by chenrui333 — APPROVED** [review]

**Comment by chenrui333** [comment]
failed on ventura intel again.

**Comment by timsutton** [comment]
Figured it out – it's lack of support for the newer Xcode / clang linker, and perhaps the Ventura x86_64 runner image was using more recent tools? I will rebase, include a couple reference GitHub issues and conditionally set the `-ld_classic` linker flag in the default `dmd.conf` `if DevelopmentTools.clang_build_version >= 1500`.

**Comment by timsutton** [comment]
On second thought – maybe that should just be a default in `dmd.conf`? Since `DevelopmentTools.clang_build_version` is set only at brew install time, while a user could later install updated Xcode tools that would break it.

**Comment by timsutton** [comment]
Maybe not.. looks like if the tools are old enough then it doesn't know that flag. Maybe we start with Ventura as the lowest OS which gets this flag? Presumably Monterey is too old to be able to run an Xcode/CLT Apple clang which would need this flag anyway. Ventura is the OS which straddles the two possibilities.

**Review by chenrui333 — COMMENTED** [review]
LGTM

File: Formula/d/dmd.rb [hunk 1]
```diff
@@ -65,9 +65,19 @@ def install
...
 
+    dflags = "-I#{opt_include}/dlang/dmd -L-L#{opt_lib}"
+    # We include the -ld_classic linker argument in dmd.conf because it seems to need
+    # changes upstream to support the newer linker:
+    # https://forum.dlang.org/thread/jwmpdecwyazcrxphttoy@forum.dlang.org?page=1
+    # https://github.com/ldc-developers/ldc/issues/4501
+    #
+    # Also, macOS can't run CLT/Xcode new enough to need this flag, so restrict to Ventura
+    # and above.
+    dflags << " -L-ld_classic" if OS.mac? && MacOS.version > :monterey
```
Comment (SMillerDev) [inline Formula/d/dmd.rb:76]: I think this needs to switch on XCode 15 or later since that's what shipped the new linker

Comment (timsutton) [reply Formula/d/dmd.rb:76]: I think I agree now that switching on Xcode 15 ( logical choice for how to configure it. Originally I was thinking that this config in `dmd.conf` could be problematic for a user if the wasn't present and they _hadn't_ installed Xcode 15 yet, and then installed it later. But that seems like it's increasingly becoming an edgecase. Plus people can just edit the file themselves.

**Review by p-linnane — APPROVED** [review]
Thanks!

**Pull request merged**